ESACCI/read-biomass-ESACCI.py

```diff
@@ -0,0 +1,217 @@
+area   = np.ma.masked_array(np.zero.rand(lat.size,lon.size))
+
+nlat = lat.size
+nlon = lon.size
+
+R    = np.float64(6371007.181)
+
+for ny in range(nlat):
+    if ny == 0:
+       dlat = abs((lat[1]-lat[0])*0.01745)
+    else:
+       dlat = abs((lat[ny]-lat[ny-1])*0.01745)
+    dy = R*dlat
+    for nx in range(nlon):
+        if nx==0:
+           dlon = abs((lon[1]-lon[0])*0.01745)
+        else:
+           dlon = abs((lon[nx]-lon[nx-1])*0.01745)
+        dx = R*math.cos(0.01745*lat[ny])*dlon
+        area[ny,nx] = dx*dy
+
+def rebin(arr, new_shape):
+    """Rebin 2D array arr to shape new_shape by averaging."""
+    shape = (new_shape[0], arr.shape[0] // new_shape[0],
+             new_shape[1], arr.shape[1] // new_shape[1])
+    return arr.reshape(shape).mean(-1).mean(1)
+
+# read single netCDF file
+filename = DataDir + '/' + sourceID + '/' + local_source
+ag=Dataset(filename,'r',format='NETCDF4')
+
+agb1 = ag.variables['agb']
+lat1 = ag.variables['lat']
+lon1 = ag.variables['lon']
+
+agb  = agb1[0,::-1,:]
+lat1 = lat1[::-1]
+
+del agb1
+
+nlat1 = lat1.size
+nlon1 = lon1.size
+
+nlat2 = int(nlat1/2)
+nlon2 = int(nlon1/2)
+
+biomass = np.ma.masked_array(np.zero.rand(nlat,nlon), dtype=float64)
+
+biomass[...] = 0.
+
+for ny in range(nlat1):
+    if ny == 0:
+       dlat = abs((lat1[1]-lat1[0])*0.01745)
+    else:
+       dlat = abs((lat1[ny]-lat1[ny-1])*0.01745)
+    dy = R*dlat
+    for nx in range(nlon1):
+        if nx==0:
+           dlon = abs((lon1[1]-lon1[0])*0.01745)
+        else:
+           dlon = abs((lon1[nx]-lon1[nx-1])*0.01745)
+        dx = R*math.cos(0.01745*lat1[ny])*dlon
+        areas = dx*dy*1.0e-4
+
+        #+++++ to generate data in 0.5x0.5 ++++
+        iy   = int(lat1[ny]*2+90.*2)
+        ix   = int(lon1[nx]*2+180.*2)
+
+        # obtain total biomass mass at each grid cell
+        biomass[iy,ix] = biomass[iy,ix] + agb[ny,nx]*areas
+
+# convert total mass (Mg) to density (Mg/m2)
+biomass[...] = biomass[...]/area[...]
+
+# convert biomass to biomass carbon 
+biomass[:,:] = biomass[:,:]/2.
+
+# convert the unit from Mg/m2 to Kg/m2
+biomass[:,:] = biomass[:,:]*1000.
+
+data[0,:,:] = biomass[:,:]
+
+with Dataset(DataDir + "biomass.nc", mode="w") as dset:
+
+    # Create netCDF dimensions
+    dset.createDimension("time",size=  t.size)
+    dset.createDimension("lat" ,size=lat.size)
+    dset.createDimension("lon" ,size=lon.size)
+    dset.createDimension("nb"  ,size=2       )
+
+    # Create netCDF variables
+    T  = dset.createVariable("time"       ,t.dtype   ,("time"     ))
+    TB = dset.createVariable("time_bounds",t.dtype   ,("time","nb"))
+    X  = dset.createVariable("lat"        ,lat.dtype ,("lat"      ))
+    XB = dset.createVariable("lat_bounds" ,lat.dtype ,("lat","nb" ))
+    Y  = dset.createVariable("lon"        ,lon.dtype ,("lon"      ))
+    YB = dset.createVariable("lon_bounds" ,lon.dtype ,("lon","nb" ))
+    D  = dset.createVariable("biomass"    ,data.dtype,("time","lat","lon"), fill_value = -999.)
```

Comment:
When creating a large netCDF variable, add the keyword argument `zlib=True`. This will use compression and make the files about 10x smaller. You only need to do this for the big arrays, like biomass here. The smaller 1D or 2D arrays don't matter so much and so I do not use it on things like `time`, `lat`, `lon`.

Reply:
Good suggestions.